docs_src/design/adr/security/0019-spiffe.md

```diff
@@ -0,0 +1,338 @@
+# Use SPIFFE/SPIRE for On-demand Secret Store Token Generation
+
+## Status
+
+** DRAFT **
+
+## Context
+
+In security-enabled EdgeX,
+there is a component called `security-secretstore-setup`
+that seeds authentication tokens for Hashicorp Vault--EdgeX's secret store--into
+directories reserved for each EdgeX microservice.
+The implementation is provided by a sub-component, `security-file-token-provider`,
+that works off of a static configuration file (`token-config.json`)
+that configures known EdgeX services,
+and an environment variable that lists additional services that require tokens.
+The token provider creates a unique token for each service
+and attaches a custom policy to each token that limits token access
+in a manner that paritions the secret store's namespace.
+
+The current solution has some problematic aspects:
+
+* These tokens have an initial TTL of one hour (1h)
+  and become invalid if not used and renewed within that time period.
+  In EdgeX Ireland, the problem is made worse by the fact that Consul
+  access token limetime is tied to the lifetime of the secret store token.
+  Prior to Ireland, it was possible to restart `security-secretstore-setup`
+  to regenerate secret store tokens and add new ones.
+  Now, with the added compilication of Consul tokens,
+  the EdgeX framework must be shut down and restarted completely
+  in order to change the list of running services.
+  It is not possible to delay the start of EdgeX services until a later time
+  (that is, greater than the default token TTL),
+  as they will not be able to connect to the EdgeX secret store
+  to obtain required secrets.
+
+* Transmission of the authentication token requires one or more shared file systems
+  between the service and `security-secretstore-setup`.
+  In the Docker implementation,
+  this shared file system is constructed by bind-mounting a host-based directory
+  to multiple containers.
+  The snap implementation is similar, utilizing a content-interface between snaps.
+  Similar functionality is near-impossible in a Kubernetes environment
+  that supports only RWO volumes:
+  the volume containing the secret can be mounted to only a single pod at a time.
+  It is simply easier to run without security enabled in Kubernetes.
+
+* The current approach cannot support distributed services
+  without an underlying distributed file system to distribute tokens,
+  such as GlusterFS, running across the participating nodes.
+
+## Decision
+
+EdgeX will create a new service, `security-spiffe-token-provider`.
+This service will be a mutual-auth TLS service that exchanges
+a [SPIFFE](https://spiffe.io/) X.509 SVID for a secret store token.
+
+The SPIFFE token provider will take theee parameters:
+
+1. An X.509 SVID used in mutual-auth TLS for the token provider
+   and the service to cross-authenticate.
+   
+2. The reqested service key.  If blank, the service key will
+   default to the service name encoded in the SVID.
```

Comment:
Why the extra parameter, why not just always use the ID from the SVID?

Reply:
For components like app-service-configurable, with SPIFFE/SPIRE we can only identify the executable.  But app-service-configurable is a chameleon -- it assumes an identity provided on the command line and gives its its own slice of the secret store and configuration provider.    It plugs a hole in that the SPIFFE/SPIRE plugin, running in a snap or as a plain process, can't introspect command line args.